File: facecat_cpp/source/sock/FCClientSockets.cpp
#include "..\\..\\stdafx.h"
#include "..\\..\\include\\sock\\FCClientSocket.h"
#include "..\\..\\include\\sock\\FCClientSockets.h"

namespace FaceCat{
	map<int, FCClientSocket*> m_clients;

	RecvMsg m_recvServerMsgCallBack = 0;

	WriteClientLog m_writeClientLogCallBack = 0;
	
	void FCClientSockets::recvServerMsg(int socketID, int localSID, const char *str, int len){
		if(m_recvServerMsgCallBack){
			m_recvServerMsgCallBack(socketID, localSID, str, len);
		}
	}

	void FCClientSockets::writeLog(int socketID, int localSID, int state, const char *log){
		if(m_writeClientLogCallBack){
			m_writeClientLogCallBack(socketID, localSID, state, log);
		}
	}

	///////////////////////////////////////////////////////////////////////////////////////////////////////////

	int FCClientSockets::close(int socketID){
		map<int, FCClientSocket*>::iterator sIter = m_clients.find(socketID);
		if(sIter != m_clients.end()){
			FCClientSocket *client = sIter->second;
			int ret = client->close(socketID);
			m_clients.erase(sIter);
			client = 0;
			return ret;
		}
		return SOCKET_ERROR;
	}

	int FCClientSockets::connect(int type, int proxyType, const char *ip, int port, const char *proxyIp, int proxyPort, const char *proxyUserName, const char *proxyUserPwd, const char *proxyDomain, int timeout){
		FCClientSocket *client = new FCClientSocket(type, proxyType, ip, port, proxyIp, proxyPort, proxyUserName, proxyUserPwd, proxyDomain, timeout);
		ConnectStatus ret = client->connect();
		if(ret != CONNECT_SERVER_FAIL){
			int socketID = (int)client->m_hSocket;
			m_clients[socketID] = client;
			client->m_writeLogEvent = writeLog;
			client->m_recvEvent = recvServerMsg;
			return socketID;
		}
		else{
			delete client;
			client = 0;
			return -1;
		}
	}

	void FCClientSockets::registerLog(WriteClientLog writeLogCallBack){
		m_writeClientLogCallBack = writeLogCallBack;
	}

	void FCClientSockets::registerRecv(RecvMsg recvMsgCallBack){
		m_recvServerMsgCallBack = recvMsgCallBack;
	}

	int FCClientSockets::send(int socketID, const char *str, int len){
		return FCClientSocket::send(socketID, str, len);
	}

	int FCClientSockets::sendTo(int socketID, const char *str, int len){
		map<int, FCClientSocket*>::iterator sIter = m_clients.find(socketID);
		if(sIter != m_clients.end()){
			FCClientSocket *client = sIter->second;
			return client->sendTo(str, len);
		}
		return -1;
	}
}

File: facecat_cs/sock/FCClientSockets.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace FaceCat {
    public class FCClientSockets {
        private static HashMap<int, FCClientSocket> m_clients = new HashMap<int, FCClientSocket>();

        private static FCSocketListener m_listener;

        private static int m_socketID;

        public static int close(int socketID) {
            int ret = -1;
            if (m_clients.containsKey(socketID)) {
                FCClientSocket client = m_clients.get(socketID);
                ret = client.close();
                m_clients.remove(socketID);
            }
            return ret;
        }

        public static int connect(int type, int proxyType, String ip, int port, String proxyIp, int proxyPort, String proxyUserName, String proxyUserPwd, String proxyDomain, int timeout) {
            FCClientSocket client = new FCClientSocket(type, (long)proxyType, ip, port, proxyIp, proxyPort, proxyUserName, proxyUserPwd, proxyDomain, timeout);
            ConnectStatus ret = client.connect();
            if (ret != ConnectStatus.CONNECT_SERVER_FAIL) {
                m_socketID++;
                int socketID = m_socketID;
                client.m_hSocket = m_socketID;
                m_clients.put(socketID, client);
                return socketID;
            }
            else {
                client.delete();
                return -1;
            }
        }

        public static void recvClientMsg(int socketID, int localSID, byte[] str, int len) {
            m_listener.callBack(socketID, localSID, str, len);
        }

        public static int send(int socketID, byte[] str, int len) {
            int ret = -1;
            if (m_clients.containsKey(socketID)) {
                FCClientSocket client = m_clients.get(socketID);
                ret = client.send(str, len);
            }
            return ret;
        }

        public static int sendTo(int socketID, byte[] str, int len) {
            int ret = -1;
            if (m_clients.containsKey(socketID)) {
                FCClientSocket client = m_clients.get(socketID);
                ret = client.sendTo(str, len);
            }
            return ret;
        }

        public static int setListener(FCSocketListener listener) {
            m_listener = listener;
            return 1;
        }

        public static void writeClientLog(int socketID, int localSID, int state, String log) {
            m_listener.writeLog(socketID, localSID, state, log);
        }
    }
}
